File: src/KrimzLib/window/key.cpp
#include "KrimzLib/window/key.h"


kl::key::key() {
	state = false;
	press = []() {};
	down = []() {};
	release = []() {};
}

kl::key::operator bool() {
	return state;
}

void kl::key::update(bool newState) {
	if (!state && newState) this->press();
	else if (state && !newState) this->release();
	state = newState;
}

File: source/kl/window/Key.java
package kl.window;

public class Key {
    public interface KeyCallback { void method(); }
    public KeyCallback press = () -> {};
    public KeyCallback down = () -> {};
    public KeyCallback release = () -> {};

    private boolean state = false;

    public Key() {}

    public boolean isDown() {
        return state;
    }

    protected void update(boolean newState) {
        if (!state && newState) {
            press.method();
        }
        else if (state && !newState) {
            release.method();
        }
        state = newState;
    }
    protected void update() {
        if (state) {
            down.method();
        }
    }
}
